sdks/java/harness/src/main/java/org/apache/beam/fn/harness/state/MultimapUserState.java

```diff
@@ -235,6 +254,122 @@ public K next() {
     };
   }
 
+  @SuppressWarnings({
+    "nullness" // TODO(https://github.com/apache/beam/issues/21068)
+  })
+  /*
+   * Returns an Iterable containing all <K, V> entries in this multimap.
+   */
+  public PrefetchableIterable<Map.Entry<K, V>> entries() {
+    checkState(
+        !isClosed,
+        "Multimap user state is no longer usable because it is closed for %s",
+        keysStateRequest.getStateKey());
+    if (isCleared) {
+      Map<Object, KV<K, List<V>>> pendingAddsNow = new HashMap<>(pendingAdds);
+      return PrefetchableIterables.concat(
+          Iterables.concat(
+              Iterables.transform(
+                  pendingAddsNow.entrySet(),
+                  entry ->
+                      Iterables.transform(
+                          entry.getValue().getValue(),
+                          value -> Maps.immutableEntry(entry.getValue().getKey(), value)))));
+    }
+
+    Set<Object> pendingRemovesNow = new HashSet<>(pendingRemoves.keySet());
+    // Make a deep copy of pendingAdds so this iterator represents a snapshot of state at the time
+    // it was created.
+    Map<K, List<V>> pendingAddsNow = new HashMap<>();
+    for (Map.Entry<Object, KV<K, List<V>>> entry : pendingAdds.entrySet()) {
+      pendingAddsNow.put(
+          entry.getValue().getKey(), new ArrayList<>()); // entry.getValue().getValue());
+      for (V value : entry.getValue().getValue()) {
+        List<V> values = pendingAddsNow.get(entry.getValue().getKey());
+        values.add(value);
+      }
+    }
+    return new PrefetchableIterables.Default<Map.Entry<K, V>>() {
+      @Override
+      public PrefetchableIterator<Map.Entry<K, V>> createIterator() {
+        return new PrefetchableIterator<Map.Entry<K, V>>() {
+          PrefetchableIterator<Map.Entry<K, V>> persistedEntriesIterator =
+              PrefetchableIterables.concat(
+                      Iterables.concat(
+                          Iterables.transform(
+                              persistedEntries,
+                              entry ->
+                                  Iterables.transform(
+                                      entry.getValue(),
+                                      value -> Maps.immutableEntry(entry.getKey(), value)))))
+                  .iterator();
+          Iterator<Map.Entry<K, V>> pendingAddsNowIterator;
+          boolean hasNext;
```

Comment:
hasNextReady? it's confusing as is since hasNext=false doesn't mean it doesn't have next, just that we haven't calculated it.

or perhaps you can remove this variable and just mark nextEntry nullable and comment that if it is non-null it is calculated next element to return.

Reply:
Sounds good. Got rid of hasNext and reset nextEntry to null when calling next().